[SPARK-22966][PYTHON][SQL] Python UDFs with returnType=StringType should treat return values of datetime.date or datetime.datetime as unconvertible

rednaxelafx opened the pull request:
## What changes were proposed in this pull request?

Perform appropriate conversions for results coming from Python UDFs that return `datetime.date` or `datetime.datetime`.

Before this PR, Pyrolite would unpickle both `datetime.date` and `datetime.datetime` into a `java.util.Calendar`, which Spark SQL doesn't understand, which then leads to incorrect results. An example of such incorrect result is:

```
>>> py_date = udf(datetime.date)
>>> spark.range(1).select(py_date(lit(2017), lit(10), lit(30))).show(truncate=False)
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|date(2017, 10, 30)                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|java.util.GregorianCalendar[time=?,areFieldsSet=false,areAllFieldsSet=false,lenient=true,zone=sun.util.calendar.ZoneInfo[id="America/Los_Angeles",offset=-28800000,dstSavings=3600000,useDaylight=true,transitions=185,lastRule=java.util.SimpleTimeZone[id=America/Los_Angeles,offset=-28800000,dstSavings=3600000,useDaylight=true,startYear=0,startMode=3,startMonth=2,startDay=8,startDayOfWeek=1,startTime=7200000,startTimeMode=0,endMode=3,endMonth=10,endDay=1,endDayOfWeek=1,endTime=7200000,endTimeMode=0]],firstDayOfWeek=1,minimalDaysInFirstWeek=1,ERA=?,YEAR=2017,MONTH=9,WEEK_OF_YEAR=?,WEEK_OF_MONTH=?,DAY_OF_MONTH=30,DAY_OF_YEAR=?,DAY_OF_WEEK=?,DAY_OF_WEEK_IN_MONTH=?,AM_PM=0,HOUR=0,HOUR_OF_DAY=0,MINUTE=0,SECOND=0,MILLISECOND=?,ZONE_OFFSET=?,DST_OFFSET=?]|
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
After this PR, the same query above would give correct results:
```
>>> spark.range(1).select(py_date(lit(2017), lit(10), lit(30))).show(truncate=False)
+------------------+
|date(2017, 10, 30)|
+------------------+
|2017-10-30        |
+------------------+
```

An explicit non-goal of this PR is to change the behavior of timezone awareness or timezone settings of `datetime.datetime` objects collected from a `DataFrame`.
Currently PySpark always returns such `datetime.datetime` objects as timezone unaware (naive) ones that respect Python's current local timezone (#19607 changed the default behavior for Pandas support but not for plain `collect()`). This PR does not change that behavior.

## How was this patch tested?

Added some unit tests to `pyspark.sql.tests` for such UDFs, so that
 * `datetime.date` -> `StringType`
 * `datetime.date` -> `DateType`
 * `datetime.datetime` -> `StringType`
 * `datetime.datetime` -> `TimestampType`
* `datetime.datetime` with non-default timezone
* `datetime.datetime` with null timezone (naive datetime)
cases are covered.

SparkQA commented:
**[Test build #85709 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85709/testReport)** for PR 20163 at commit [`ca026d3`](https://github.com/apache/spark/commit/ca026d31a489f1e0eb451fe85df97083659d0f67).

cloud-fan commented on python/pyspark/sql/udf.py:
```diff
@@ -26,6 +26,28 @@
+    # because the format of the string should be different, depending on the type of the input
+    # object. So for those two specific types we eagerly convert them to string here, where the
+    # Python type information is still intact.
+    if returnType == StringType():
```
This is to handle when a python udf returns `date` or `datetime` but mark the return type as string?

viirya replied:
I have a question, why we need to handle this type conversion? If we expect correct string format, isn't it more reasonable to convert the date/datetime to strings in the udf, instead of adding this conversion implicitly?

cloud-fan commented:
LGTM,  cc @ueshin @icexelloss does this behavior consistent with pandas UDF?

icexelloss commented:
I think Scalar and Group map UDF expect pandas Series of datetime64[ns] (native pandas timestamp type) instead of a pandas Series of datetime.date and datetime.datetime object. I don't think it's necessary to have pandas UDF to work with a pandas Series of datetime.date or datetime.datetime object, as the standard type of timestamp is datetime64[ns] in pandas.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvaluatePython.scala:
```diff
@@ -120,10 +121,18 @@ object EvaluatePython {
     case (c: java.math.BigDecimal, dt: DecimalType) => Decimal(c, dt.precision, dt.scale)
 
     case (c: Int, DateType) => c
```
BTW, as a side note, I think we can make the converter for the type and then reuse it.

HyukjinKwon replied:
Of course, separate change obviously.

HyukjinKwon commented on python/pyspark/sql/udf.py:
```diff
@@ -26,6 +26,28 @@
+    def coerce_to_str(v):
+        import datetime
+        if type(v) == datetime.date or type(v) == datetime.datetime:
+            return str(v)
```
I think it's weird that we have a cast here alone ... Can't we register a custom Pyrolite unpickler? Does it make the things more complicated?

HyukjinKwon commented:
Wait .. Isn't this because we failed to call `toInternal` by the return type? Please give me few days .. will double check tonight.

SparkQA commented:
**[Test build #85709 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/85709/testReport)** for PR 20163 at commit [`ca026d3`](https://github.com/apache/spark/commit/ca026d31a489f1e0eb451fe85df97083659d0f67).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/85709/
Test PASSed.

HyukjinKwon commented:
The problem here seems, `returnType` is mismatched to the value. In case of `DateType`, it needs an explicit conversion into integers:

https://github.com/apache/spark/blob/1c9f95cb771ac78775a77edd1abfeb2d8ae2a124/python/pyspark/sql/types.py#L170-L171

https://github.com/apache/spark/blob/1c9f95cb771ac78775a77edd1abfeb2d8ae2a124/python/pyspark/sql/types.py#L173-L175

which will be called via in `worker.py`

https://github.com/apache/spark/blob/64817c423c0d82a805abd69a3e166e5bfd79c739/python/pyspark/worker.py#L70-L74

If the `returnType` is `StringType`, then it doesn't need the conversion because Pyrolite and serialization work fine between them up to my knowledge:

https://github.com/apache/spark/blob/1c9f95cb771ac78775a77edd1abfeb2d8ae2a124/python/pyspark/sql/types.py#L141-L145

https://github.com/apache/spark/blob/1c9f95cb771ac78775a77edd1abfeb2d8ae2a124/python/pyspark/sql/types.py#L76-L82


So, here:


https://github.com/apache/spark/blob/64817c423c0d82a805abd69a3e166e5bfd79c739/python/pyspark/worker.py#L70-L74


we will send the return values as are without conversion, which ends up with `datetime.date` -> `java.util.Calendar` as you described in the PR description. Therefore, I don't think the current fix in `EvaluatePython.scala` is reachable in the reproducer above.

For the fix in Python side in `udf.py`, this is a band-aid fix. To deal with this problem correctly, I believe we should do something like:

```diff
diff --git a/python/pyspark/sql/types.py b/python/pyspark/sql/types.py
index 146e673ae97..37137e02c08 100644
--- a/python/pyspark/sql/types.py
+++ b/python/pyspark/sql/types.py
@@ -144,6 +144,17 @@ class StringType(AtomicType):

     __metaclass__ = DataTypeSingleton

+    def needConversion(self):
+        return True
+
+    def toInternal(self, v):
+        if v is not None:
+            return str(v)
+
+    def fromInternal(self, v):
+        if v is not None:
+            return str(v)
+
```

but then this will bring performance regression because `str` is required to be called every value. This extra function call could cause performance regression, for example, see both https://github.com/apache/spark/pull/19246 and https://github.com/apache/spark/pull/19249.

I am less sure this is something we should allow. Can we simply document this saying `returnType` should be compatible with the actual return value?

icexelloss commented:
I ran some experiments:
```
py_date = udf(datetime.date, DateType())
py_timestamp = udf(datetime.datetime, TimestampType())
```
This works correctly
```
spark.range(1).select(py_date(lit(2017), lit(10), lit(30))).show()
spark.range(1).select(py_timestamp(lit(2017), lit(10), lit(30))).show()
```
Result:
```
+------------------+
|date(2017, 10, 30)|
+------------------+
|        2017-10-30|
+------------------+

+----------------------+
|datetime(2017, 10, 30)|
+----------------------+
|   2017-10-30 00:00:00|
+----------------------+
```

The change that the PR proposes seem to be coercing python `datetime.datetime` and `datetime.date` to the python datetime string representation rather the java one. We could call function `str` on the return value of the python udf if it's a String type to get the python string representation, but this probably needs some microbenchmark to see the performance implication.

rednaxelafx commented:
Thanks for all of your comments, @HyukjinKwon and @icexelloss !
I'd like to wait for more discussions / suggestions on whether or not we want a behavior change that makes this reproducer work, or a simple document change that'll just say PySpark doesn't support mismatching `returnType`.

All of what you guys mentioned are correct. Sorry for the mess, I actually got myself confused...
It's been a while since I first noticed the problem and came up with a patch, and obviously when I picked it back up I've lost some context as to what exactly failed in the first place...

Both @HyukjinKwon and @icexelloss correctly pointed out that the bug only happens when the `udf()` creation declared a mismatching type versus what it actually returns. In the reproducer, the declared UDF return type is the default `string` type but the actual return types were `datetime.date` / `datetime.datetime`. That followed the path of not going through in `pyspark/sql/types.py`, so it went through to Pyrolite getting unpickled as a `java.util.Calendar`.

A note on how I got here: the reason why my current PR (incorrectly) contained the cases for `case (Calendar, DateType)` and friends was that, initially I only had a reproducer for a Python UDF actually returning `datetime.date` but was using the default return type (`string`), and I had fixed it by introducing a case for converting from `java.util.Calendar` to the appropriate type in `EvaluatePython.scala`. At that time that case was certainly executed and it did give me correct results for that single reproducer. I did notice that the cases where the UDF return type was correctly declared was working correctly.

But then I realized I also needed to handle the case where I have to tell apart `datetime.date` and `datetime.datetime`, and then I can't do that in a single `case (Calendar, StringType)` in `EvaluatePython.scala` anymore because I'm lacking the type information from the Python side at that point. So I went back to the Python side and thought I needed to handle `datetime.date` / `datetime.datetime` separately, but eventually they ended up both being handled by just a `str(value)` coercion in a band-aid fix in `udf.py`. The version that @HyukjinKwon suggested above is indeed a proper version of that.
At this point the new code in `EvaluatePython.scala` and friends are dead code.

To address a point from @icexelloss :
> The change that the PR proposes seem to be coercing python datetime.datetime and datetime.date to the python datetime string representation rather the java one.
The reason why I used `str()` there was that both Python and Spark SQL followed the same default string formats for date (`yyyy-MM-dd`) and datetime (`yyyy-MM-dd HH:mm:ss`), e.g.
```c
static PyObject *
date_isoformat(PyDateTime_Date *self)
{
    return PyUnicode_FromFormat("%04d-%02d-%02d",
                                GET_YEAR(self), GET_MONTH(self), GET_DAY(self));
}
```
and
```scala
  // `SimpleDateFormat` is not thread-safe.
  private val threadLocalDateFormat = new ThreadLocal[DateFormat] {
    override def initialValue(): SimpleDateFormat = {
      new SimpleDateFormat("yyyy-MM-dd", Locale.US)
    }
  }
```

HyukjinKwon commented:
Hey @rednaxelafx that's fine. We all make mistake and I usually think it's always better then not trying. I also made a mistake at the first time. It was easier to debug this with your comments and details in the PR description. Thank you.

> I'd like to wait for more discussions / suggestions on whether or not we want a behavior change that makes this reproducer work, or a simple document change that'll just say PySpark doesn't support mismatching returnType.

So, few options might be ...

1. Simple document this

2. `str` logics in `type.StringType` - in this case, I think we should do a small banchmark. It'd would be so hard and I think you could reuse commands I used here - https://github.com/apache/spark/pull/19246#discussion_r139874732

3. Investigate the way to register a custom Pyrolite unpickler that converts `datetime.date*` to `Timestamp` or `Date`. I believe we already have some custom fixes there.

HyukjinKwon commented:
@ueshin @icexelloss @cloud-fan @rednaxelafx, which one would you prefer?

To me, I like 1 at most. If the perf diff is trivial, 2. is also fine. If 3. works fine, I think I am also fine with it.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala:
```diff
@@ -193,6 +193,24 @@ object DateTimeUtils {
+  }
+
+  /**
+   * Returns SQLTimestamp from java.util.Calendar (microseconds since epoch)
```
(Matching the comment of fromJavaCalendarForDate.)
nit: Returns the number of microseconds since epoch from java.util.Calendar.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvaluatePython.scala:
```diff
@@ -120,10 +121,18 @@ object EvaluatePython {
 
     case (c: Int, DateType) => c
+    // Pyrolite will unpickle a Python datetime.date to a java.util.Calendar
+    case (c: Calendar, DateType) => DateTimeUtils.fromJavaCalendarForDate(c)
```
so we will never hit this?

cloud-fan replied:
seems what we need is a `case (c: Calendar, StringType) => ...`

HyukjinKwon replied:
I think he did this in Python side because here we don't know if `Calendar` is from `datetime.date` or `datetime.datetime`.

HyukjinKwon commented:
@cloud-fan, actually I have the similar question too - https://github.com/apache/spark/pull/20163#discussion_r160017637. I tend to agree with it and I think we disallow this and document this.

Just want to check if you feel strongly about this. If we need to support this, I believe the ways are 2. or 3. in https://github.com/apache/spark/pull/20163#issuecomment-355717642.

cloud-fan commented:
The current behavior looks weird, we should either throw exception and ask users to give a corrected return type or fix it via proposal 2.

HyukjinKwon commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvaluatePython.scala:
```diff
@@ -120,10 +121,18 @@ object EvaluatePython {
 
     case (c: Int, DateType) => c
+    // Pyrolite will unpickle a Python datetime.date to a java.util.Calendar
+    case (c: Calendar, DateType) => DateTimeUtils.fromJavaCalendarForDate(c)
```
How about we return `null` in this case? Other cases seems also returning `null` if it fails to be converted:

```
>>> from pyspark.sql.functions import udf
>>> f = udf(lambda x: x, "double")
>>> spark.range(1).select(f("id")).show()
+------------+
|<lambda>(id)|
+------------+
|        null|
+------------+
```

Seems we can do it like:

```scala

    case StringType => (obj: Any) => nullSafeConvert(obj) {
      case c: Calendar => null
      case _ => UTF8String.fromString(obj.toString)
    }
```

cloud-fan replied:
Yea it's consistent with other un-convertible cases, but `StringType` is the default return type, I'm afraid many users many hit this and get confused.

HyukjinKwon replied:
Right. Let's go ahead for 2. then. I am fine if it's done as an exception for practical purpose. Maybe we could add an if `isinstance(.., basestring)` and return directly as a shortcut. I haven't checked the perf diff but I think we can do it easily via profile as I mentioned above.

HyukjinKwon replied:
WDYT about ^ @ueshin?

ueshin replied:
Yeah, 2. should work for `StringType`.

I'd also like to add some documents like 1. for users to be careful about the return type. I've found that `udf`s return `null` and `pandas_udf`s throw some exception in most case when the return type is mismatching.
Of course we can try to make the behavior differences between `udf` and `pandas_udf` closer as possible in the future, but I think it is the best effort basis for the mismatching return type.

icexelloss replied:
My 2 cents:
* I am +0 for 2. I think having type coercion for str type but not for other types can be confusing to users. Realistically, I think any udf users would hit the `null` result for mismatch type anyway (I have hit it so many times..) and will learn that `null` means type mismatch. Even we make the behavior for str a bit friendly, they will likely to hit the issue with other types anyway. I don't think "default returnType for udf is str" is a strong reason for "having special type coercion for str", they seem orthogonal. 
**I prefer that we keep the type mismatch behavior consistent for str types vs other types (return null) and document this more clearly.**
* I find returning `null` for type mismatch is an unintuitive behavior of row-at-a-time udf and prefer not to replicate it in pandas udf.

ueshin commented:
I investigated the behavior differences between `udf` and `pandas_udf` for the wrong return types and found there are many differences actually.
Basically `udf`s return `null` as @HyukjinKwon mentioned, whereas `pandas_udf`s throw some `ArrowException`. There seem some exceptions, though.

HyukjinKwon commented:
Probably we consider to catch and set nulls in pandas_udf if possible to match the behaviour with udf ...

rednaxelafx commented:
Given the above discussion, do we have consensus on all of the following:
- Update the documentation for PySpark UDFs to warn about the behavior of mismatched declared `returnType` vs actual runtime return values
- Make Python UDFs that declared `returnType` as `StringType` recognize `java.util.Calendar` and convert the value into a `null` (as in the example from @HyukjinKwon ), essentially marking it as unconvertible.

I believe we all agree on the first point. The second point above is in line with @icexelloss 's opinion, which I tend to agree in terms of API semantic consistency. It might not be as user-friendly as Option 2 from @HyukjinKwon , but it's less magic and more consistent. I tend to find more consistency leads to less surprises.

If we have consensus then I'll update the JIRA ticket and this PR to reflect that.

cloud-fan commented:
SGTM

HyukjinKwon commented:
One more SGTM

viirya commented:
+1

icexelloss commented:
+1

Commit:
[SPARK-22966][PYTHON][SQL] Python UDFs with returnType=StringType should treat return values of datetime.date or datetime.datetime as unconvertible

Add conversion to PySpark to mark Python UDFs that declared returnType=StringType() but actually returned a datatime.date or datetime.datetime as unconvertible, i.e. converting it to null.

Also added a new unit test to pyspark/sql/tests.py to reflect current semantics of Python UDFs returning a value of mismatched type with the declared returnType.

SparkQA commented:
**[Test build #86000 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86000/testReport)** for PR 20163 at commit [`4c7bcc1`](https://github.com/apache/spark/commit/4c7bcc1c8687d9d3c39a0ac8b2f03692d699b948).

SparkQA commented:
**[Test build #86000 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86000/testReport)** for PR 20163 at commit [`4c7bcc1`](https://github.com/apache/spark/commit/4c7bcc1c8687d9d3c39a0ac8b2f03692d699b948).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86000/
Test FAILed.

SparkQA commented:
**[Test build #86003 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86003/testReport)** for PR 20163 at commit [`d307cee`](https://github.com/apache/spark/commit/d307ceeaa581dc219ae744482a6d3c374b3a3025).

SparkQA commented:
**[Test build #86003 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86003/testReport)** for PR 20163 at commit [`d307cee`](https://github.com/apache/spark/commit/d307ceeaa581dc219ae744482a6d3c374b3a3025).
 * This patch **fails from timeout after a configured wait of \`250m\`**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86003/
Test FAILed.

rednaxelafx commented:
jenkins retest this please

SparkQA commented:
**[Test build #86023 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86023/testReport)** for PR 20163 at commit [`d307cee`](https://github.com/apache/spark/commit/d307ceeaa581dc219ae744482a6d3c374b3a3025).

SparkQA commented:
**[Test build #86023 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/86023/testReport)** for PR 20163 at commit [`d307cee`](https://github.com/apache/spark/commit/d307ceeaa581dc219ae744482a6d3c374b3a3025).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/86023/
Test FAILed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvaluatePython.scala:
```diff
@@ -144,6 +145,7 @@ object EvaluatePython {
 
     case StringType => (obj: Any) => nullSafeConvert(obj) {
+      case _: Calendar => null
       case _ => UTF8String.fromString(obj.toString)
```
shall we blacklist more types? e.g. if a udf returns decimal and mark the return type as string type, is it a mismatch?

rednaxelafx replied:
I was pounding on that yesterday, too... somehow I have this feeling that no matter which direction we take, there's no good answer to type mismatch situations.

Let's say if we blacklist more types, should we document the list so that it's clear what will definitely NOT work?

HyukjinKwon replied:
For the perfectness, I think we should check all the types, https://github.com/irmen/Pyrolite,

```
PYTHON    ---->     JAVA
------              ----
None                null
bool                boolean
int                 int
long                long or BigInteger  (depending on size)
string              String
unicode             String
complex             net.razorvine.pickle.objects.ComplexNumber
datetime.date       java.util.Calendar
datetime.datetime   java.util.Calendar
datetime.time       net.razorvine.pickle.objects.Time
datetime.timedelta  net.razorvine.pickle.objects.TimeDelta
float               double   (float isn't used) 
array.array         array of appropriate primitive type (char, int, short, long, float, double)
list                java.util.List<Object>
tuple               Object[]
set                 java.util.Set
dict                java.util.Map
bytes               byte[]
bytearray           byte[]
decimal             BigDecimal    
custom class        Map<String, Object>  (dict with class attributes including its name in "__class__")
Pyro4.core.URI      net.razorvine.pyro.PyroURI
Pyro4.core.Proxy    net.razorvine.pyro.PyroProxy
Pyro4.errors.*      net.razorvine.pyro.PyroException
Pyro4.utils.flame.FlameBuiltin     net.razorvine.pyro.FlameBuiltin 
Pyro4.utils.flame.FlameModule      net.razorvine.pyro.FlameModule 
Pyro4.utils.flame.RemoteInteractiveConsole    net.razorvine.pyro.FlameRemoteConsole 
```

and then check if the string conversion looks reasonably consistent by `obj.toString`. If not, we add it in the blacklist.

Another possibility is to whitelist `String`, but then I guess this is rather a radical change.

cloud-fan replied:
> check if the string conversion looks reasonably consistent by obj.toString. If not, we add it in the blacklist.

hmm, this seems weird as the type mismatch now is defined by Pyrolite object's `toString` behavior...

HyukjinKwon replied:
So, for now .. I think it's fine as a small fix as is ... We are going to document that the return type and return value should be matched anyway ..

So, expected return values will be:

```python
# Mapping Python types to Spark SQL DataType
_type_mappings = {
    type(None): NullType,
    bool: BooleanType,
    int: LongType,
    float: DoubleType,
    str: StringType,
    bytearray: BinaryType,
    decimal.Decimal: DecimalType,
    datetime.date: DateType,
    datetime.datetime: TimestampType,
    datetime.time: TimestampType,
}
```

Seems, we can also check if the string conversion looks reasonable and then blacklist `net.razorvine.pickle.objects.Time` if not ...

How does this sound to you @cloud-fan and @rednaxelafx?

HyukjinKwon replied:
Oh, I didn't see the comment above when I write my comment.

HyukjinKwon replied:
BTW, seems there is another hole when we actually do the internal conversion with unexpected types:

```python
>>> from pyspark.sql.functions import udf
>>> f = udf(lambda x: x, "date")
>>> spark.range(1).select(f("id")).show()
```

```
org.apache.spark.api.python.PythonException: Traceback (most recent call last):
  File "./python/pyspark/worker.py", line 229, in main
    process()
  File "./python/pyspark/worker.py", line 224, in process
    serializer.dump_stream(func(split_index, iterator), outfile)
  File "./python/pyspark/worker.py", line 149, in <lambda>
    func = lambda _, it: map(mapper, it)
  File "<string>", line 1, in <lambda>
  File "./python/pyspark/worker.py", line 72, in <lambda>
    return lambda *a: toInternal(f(*a))
  File "/.../pyspark/sql/types.py", line 175, in toInternal
    return d.toordinal() - self.EPOCH_ORDINAL
AttributeError: 'int' object has no attribute 'toordinal'
```

HyukjinKwon replied:
I think there is no perfect solution ..  I think https://github.com/apache/spark/pull/20163#discussion_r161363004 sounds good enough as a fix for this issue for now ..

HyukjinKwon replied:
@cloud-fan, how about something like this then?

```scala
    case StringType => (obj: Any) => nullSafeConvert(obj) {
      // Shortcut for string conversion
      case c: String => UTF8String.fromString(c)

      // Here, we return null for 'array', 'tuple', 'dict', 'list', 'datetime.datetime',
      // 'datetime.date' and 'datetime.time' because those string conversions are
      // not quite consistent with SQL string representation of data.
      case _: java.util.Calendar | _: net.razorvine.pickle.objects.Time |
           _: java.util.List[_] | _: java.util.Map[_, _] =>
        null
      case c if c.getClass.isArray => null

      // Here, we keep the string conversion fall back for compatibility.
      // TODO: We should revisit this and rewrite the type conversion logic in Spark 3.x.
      case other => UTF8String.fromString(other.toString)
    }
```

My few tests:

`datetime.time`:

```
from pyspark.sql.functions import udf
from datetime import time

f = udf(lambda x: time(0, 0), "string")
spark.range(1).select(f("id")).show()
```

```
+--------------------+
|        <lambda>(id)|
+--------------------+
|Time: 0 hours, 0 ...|
+--------------------+
```

`array`:

```
from pyspark.sql.functions import udf
import array

f = udf(lambda x: array.array("c", "aaa"), "string")
spark.range(1).select(f("id")).show()
```

```
+------------+
|<lambda>(id)|
+------------+
| [C@11618d9e|
+------------+
```

`tuple`:

```
from pyspark.sql.functions import udf

f = udf(lambda x: (x,), "string")
spark.range(1).select(f("id")).show()
```

```
+--------------------+
|        <lambda>(id)|
+--------------------+
|[Ljava.lang.Objec...|
+--------------------+
```

`list`:

```
from pyspark.sql.functions import udf
from datetime import datetime

f = udf(lambda x: [datetime(1990, 1, 1)], "string")
spark.range(1).select(f("id")).show()
```

```
+--------------------+
|        <lambda>(id)|
+--------------------+
|[java.util.Gregor...|
+--------------------+
```

cloud-fan replied:
looks good

felixcheung replied:
btw, the array case seems a bit weird?

HyukjinKwon commented:
Let's leave this closed.
